Gallery-Server/src/main/java/edu/yu/cs/gallery/Utility.java

```diff
@@ -22,35 +23,77 @@
     }
     
-    public Response redirect (long[] galleries, @Context UriInfo uriInfo) throws URISyntaxException {
+    //URI must contain the string "batch" in place of the galleryId 
+    @SuppressWarnings("all")
+    private int redirect(long galleryId, UriInfo uriInfo, HttpMethod httpMethod, Object body, List<Object> returnEntities) throws URISyntaxException {
+        WebClient webClient = WebClient.create(this.allServers.get(galleryId).toString());
+        
+        String path = uriInfo.getPath();
+        path = path.replace("batch", Long.toString(galleryId));
+        
+        ResponseEntity<Object> response = webClient.method(httpMethod)
+                .uri(path)
+                .contentType(MediaType.APPLICATION_JSON)
+                .bodyValue(body)
+                .retrieve()
+                .toEntity(Object.class)
+                .block();
+
+        Object responseObject = response.getBody();
+        if (responseObject instanceof Collection) {
+            returnEntities.addAll((Collection)responseObject);
+        } else {
+            returnEntities.add(responseObject);
+        }
+        return response.getStatusCodeValue();
+    }
+    
+    protected Response readRedirect (long[] galleryIDs, UriInfo uriInfo, Request request) throws URISyntaxException {
         List<Object> returnEntities = new ArrayList<>();
-        for (long lg : galleries) {
-
-            WebClient webClient = WebClient.create(this.allServers.get(lg).toString());
-            String path = uriInfo.getPath();
-            path = path.replace("batch", Long.toString(lg));
-            Object response = webClient.get()
-                    .uri(path)
-                    .retrieve()
-                    .bodyToMono(Object.class).block();
-            
-            if (response instanceof Collection) {
-                returnEntities.addAll((Collection)response);
-            } else {
-                returnEntities.add(response);
+        
+        for (long galleryID : galleryIDs) {
+            if (allServers.containsKey(galleryID)) {
+                int statusCode = this.redirect(galleryID, uriInfo, HttpMethod.valueOf(request.getMethod()), "", returnEntities);
+                if (statusCode != 200) {
+                    return Response.status(statusCode).entity(returnEntities).build();
+                }
             }
         }
+        
         return Response.status(Status.OK).entity(returnEntities).build();
     }
+    
+    protected Response writeRedirect(Gallery[] galleries, UriInfo uriInfo, Request request) throws URISyntaxException {
+        //If the user attempts to make a batch write to a server that is not the leader, return a temporaryRedirect to the correct server.
+        if (this.gallery.id != leaderID) {
+            return temporaryRedirect(leaderID, uriInfo);
+        }
+
+        List<Object> returnEntities = new ArrayList<>();
+        
+        for (Gallery gallery : galleries) {
```

Comment:
two comments here:

* Looks like you are taking a fail fast approach. That is: if one gallery has a non success, it wont try to write to the other galleries. Correct? That is a viable approach, just want to make sure that is intentional?
* What if the current server handles the gallery in question? Does that mean the server will be making a web call to itself? Eek....

Reply:
1) Yes, we have gone for a fail-fast approach, but all the modified entities prior to the fail will be returned to the user so they can see what was changed.

2) The logic has been reworked to address this. Now, a lambda expression is passed in which contains overrides .run() to excecute the local method corresponding to what an HTTP call would have been. For examople, a batch GET call which includes gallery of ID 1 and previously would have made an HTTP call to http://.../galleries/1 will now have a lambda passed in that executes the method getById with the param of the galleryId that is on this server.